lib/src/repo.rs

```diff
@@ -385,6 +417,31 @@ impl StoreFactories {
             .expect("Unexpected op_store type");
         op_store_factory(store_path)
     }
+
+    pub fn add_op_heads_store(&mut self, name: &str, factory: OpHeadsStoreFactory) {
+        self.op_heads_store_factories
+            .insert(name.to_string(), factory);
+    }
+
+    pub fn load_op_heads_store(&self, store_path: &Path) -> Box<dyn OpHeadsStore> {
+        let op_heads_store_type = match fs::read_to_string(store_path.join("type")) {
+            Ok(content) => content,
+            Err(err) if err.kind() == ErrorKind::NotFound => {
+                // For compatibility with existing repos. TODO: Delete in spring of 2024 or so.
```

Comment:
FYI, I have been thinking that it would be enough to keep upgrades from one minor version to the next. In that case, we could remove this code in 0.8.0. That way users can upgrade current repos (0.6.x) by running 0.7.x and then 0.8.x. I haven't asked what others thinks about that policy.

Btw, we don't really make any effort to make repos written by a newer version readable by older versions. We should *at least* make sure we add some way of making older clients error out with something like "sorry, you need at least jj 0.12 to access this repo". That's made more complicated by pluggable storage implementations, because we might have to add that functionality to each store trait (so the error might actually be more like "sorry, you need at least version 0.12 of the 'git' commit backend to use this repo", where the version number is specific to the backend).

Reply:
Adjusted this comment (and the one for op_store too)